### ISSUE TITLE
RFE: Higher level interface to portable services

### ISSUE THREAD

**Issue by AlbanBedel** [systemd/systemd#34870]
### Component

portablectl

### Is your feature request related to a problem? Please describe

For an image based embedded distro I would like to be able to deliver additional applications as portable services. Currently this can only be (easily) automated for simple portable services that only use a single prefix, no extension images and the default profile. If any of these additional features is needed there is currently no standardized way to pass this information. 

In other words portablectl currently deal with images, but I would like to be able to deal with portable services; without having to care about how many images they use, which prefixes should be enabled by default, etc.

### Describe the solution you'd like

This could be achieved by defining a config file, let say `.portable` that would contains all the information needed for presentation, attaching and running the service. Something along the lines of:

```
[Portable]
Name=My portable service
Description=This great service can be run at once
Version=8.7

RootImage=base-runtime_1.5.raw
ExtensionImages=common-lib_2.7.raw other-lib_0.4.raw my-service_8.7.raw
Prefixes=common-lib-daemon my-service
Profile=strict
```

Ideally portablectl would then show a service named after this config file and not after the root image.

### Describe alternatives you've considered

Implementing this outside of systemd.

**Comment by bluca** [comment]
> If any of these additional features is needed there is currently no standardized way to pass this information

There is? There's a fully-fledged D-Bus interface that orchestrators can use: https://www.freedesktop.org/software/systemd/man/latest/org.freedesktop.portable1.html
Why can't that be used?

**Comment by AlbanBedel** [comment]
Poor wording from my side, what I mean is the delivery of the application to the target machine. Once I have installed all my images on the target machine I still need to find out which image is the base image, which are the extensions, in which order they should be stacked, etc. Only once I have all this can I then call AttachImageWithExtensions.

Another way to put it that you currently need some script or additional service to enable a portable service. I think it would be desirable to only need some configuration file to achieve that.

**Comment by bluca** [comment]
Sorry I do not follow, you still need to do all of that regardless? You need to provide all that information, what form it takes is irrelevant. On an automated system, this is the job of the local client of the orchestrator, whatever that might be.

**Comment by AlbanBedel** [comment]
You always need this information for non trivial portable services. Without it you just have a bunch of useless images. Why should `portablectl attach NAME` only be possible with trivial images?

**Comment by bluca** [comment]
Sorry, I still do not follow - it is possible, and works just fine, I linked the API above

**Comment by AlbanBedel** [comment]
Yes, the API is fine, but the tooling could be improved. I find that the current user interface of portablectl is too complex to make the use of layered images practicable. Users should be able to just run `portablectl attach NAME` without having to care about how many images the service developer decided to use.

**Comment by bluca** [comment]
That's just not possible, if you need to use multiple images, then you need to specify them, they can't be just guessed out of thin air. Also I'm not really convinced it's a real problem, what's the issue exactly? If you can specify one, surely you can specify two?

**Comment by AlbanBedel** [comment]
If one seriously starts to use layering then there will quickly be more than 2 images. And yes you can't guess that, that's why I think it would be valuable to have a common format for that.

Currently it feel like if the dynamic linker required me to manually input every required library when I start an executable. Why should I care about such details as an end user?

**Comment by karimfromjordan** [comment]
I just came across this issue and I agree. My current workaround is to create a small bash script that handles this but it would be nice if that wasn't necessary.
